sky/serve/autoscalers.py

```diff
@@ -132,93 +131,116 @@ def collect_request_information(
-                    f'min_replicas: {self.min_replicas}, '
-                    f'max_replicas: {self.max_replicas}')
-
-        num_replicas_delta = 0
-        if num_replicas < self.min_replicas:
-            num_replicas_delta = self.min_replicas - num_replicas
-            self.upscale_counter = 0
-        elif num_replicas > self.max_replicas:
-            num_replicas_delta = self.max_replicas - num_replicas
-            self.downscale_counter = 0
-        elif target_num_replicas > num_replicas:
+        if target_num_replicas > self.target_num_replicas:
             self.upscale_counter += 1
             self.downscale_counter = 0
             if self.upscale_counter >= self.scale_up_consecutive_periods:
                 self.upscale_counter = 0
-                num_replicas_delta = target_num_replicas - num_replicas
-        elif target_num_replicas < num_replicas:
+                return target_num_replicas
+        elif target_num_replicas < self.target_num_replicas:
             self.downscale_counter += 1
             self.upscale_counter = 0
             if self.downscale_counter >= self.scale_down_consecutive_periods:
                 self.downscale_counter = 0
-                num_replicas_delta = target_num_replicas - num_replicas
+                return target_num_replicas
         else:
             self.upscale_counter = self.downscale_counter = 0
-        logger.info(f'Upscale counter: {self.upscale_counter}/'
-                    f'{self.scale_up_consecutive_periods}. '
-                    f'Downscale counter: {self.downscale_counter}/'
-                    f'{self.scale_down_consecutive_periods}')
+        return self.target_num_replicas
+
+    def _get_spot_resources_override_dict(self) -> Dict[str, Any]:
+        return {'use_spot': True, 'spot_recovery': None}
+
+    def _get_on_demand_resources_override_dict(self) -> Dict[str, Any]:
+        return {'use_spot': False, 'spot_recovery': None}
+
+    def _get_resources_override_dict(self, use_spot: bool) -> Dict[str, Any]:
+        return (self._get_spot_resources_override_dict()
+                if use_spot else self._get_on_demand_resources_override_dict())
 
-        if num_replicas_delta == 0:
+    def evaluate_scaling(
+        self,
+        replica_infos: List['replica_managers.ReplicaInfo'],
+    ) -> List[AutoscalerDecision]:
+
+        # TODO(MaoZiming): Consider non-alive replicas when auto_restart = False
+        alive_replica_infos = [info for info in replica_infos if info.is_alive]
+        num_alive_replicas = len(alive_replica_infos)
+        use_spot_list = [
+            1 if info.is_spot else 0 for info in alive_replica_infos
+        ]
+        # TODO(MaoZiming): Support mix of on-demand and spot on master.
+        assert len(use_spot_list) in [
+            0, num_alive_replicas
+        ], 'replicas should be either all spot or all on-demand'
+        use_spot = sum(use_spot_list) == num_alive_replicas
```

Comment:
Do we really need these? Also, iiuc `len(use_spot_list)` is always equal to `num_alive_replicas`?

Reply:
If the user launches all spot replicas, `len(use_spot_list)` equals `num_alive_replicas`.
If the user launches all on-demand replicas, `len(use_spot_list)` equals 0

Reply:
ditto.
this PR does not differentiate between spot and on-demand